core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -0,0 +1,438 @@
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.protocol.Errors
+import org.apache.kafka.common.record.{FileRecords, MemoryRecords, SimpleRecord}
+import org.apache.kafka.common.requests.ProduceResponse.PartitionResponse
+import org.apache.kafka.common.utils.{Time, Utils}
+
+import scala.collection.mutable
+import scala.collection.JavaConverters._
+
+
+object TransactionManager {
+  // default transaction management config values
+  val DefaultTransactionalIdExpirationMs = TimeUnit.DAYS.toMillis(7).toInt
+  val DefaultTransactionsMaxTimeoutMs = TimeUnit.MINUTES.toMillis(15).toInt
+}
+
+/**
+ * Transaction manager is part of the transaction coordinator, it manages:
+ *
+ * 1. the transaction log, which is a special internal topic.
+ * 2. the transaction metadata including its ongoing transaction status.
+ * 3. the background expiration of the transaction as well as the transactional id.
+ */
+class TransactionStateManager(brokerId: Int,
+                              zkUtils: ZkUtils,
+                              scheduler: Scheduler,
+                              replicaManager: ReplicaManager,
+                              config: TransactionConfig,
+                              time: Time) extends Logging {
+
+  this.logIdent = "[Transaction Log Manager " + brokerId + "]: "
+
+  /** shutting down flag */
+  private val shuttingDown = new AtomicBoolean(false)
+
+  /** lock protecting access to loading and owned partition sets */
+  private val partitionLock = new ReentrantLock()
+
+  /** partitions of transaction topic that are assigned to this manager, partition lock should be called BEFORE accessing this set */
+  private val ownedPartitions: mutable.Map[Int, Int] = mutable.Map()
+
+  /** partitions of transaction topic that are being loaded, partition lock should be called BEFORE accessing this set */
+  private val loadingPartitions: mutable.Set[Int] = mutable.Set()
+
+  /** transaction metadata cache indexed by transactional id */
+  private val transactionMetadataCache = new Pool[String, TransactionMetadata]
+
+  /** number of partitions for the transaction log topic */
+  private val transactionTopicPartitionCount = getTransactionTopicPartitionCount
+
+  def enablePidExpiration() {
+    scheduler.startup()
+
+    // TODO: add transaction and pid expiration logic
+  }
+
+  /**
+   * Get the transaction metadata associated with the given transactional id, or null if not found
+   */
+  def getTransaction(transactionalId: String): Option[TransactionMetadata] = {
+    Option(transactionMetadataCache.get(transactionalId))
+  }
+
+  /**
+   * Add a new transaction metadata, or retrieve the metadata if it already exists with the associated transactional id
+   */
+  def addTransaction(transactionalId: String, txnMetadata: TransactionMetadata): TransactionMetadata = {
+    val currentTxnMetadata = transactionMetadataCache.putIfNotExists(transactionalId, txnMetadata)
+    if (currentTxnMetadata != null) {
```

Comment:
Yes - this was initially missing in the `initPidRequest` i've updated it so it does add it to the log